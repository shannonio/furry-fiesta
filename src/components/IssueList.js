import React, { useEffect } from 'react';
import placeholderAvatar from '../images/person.jpeg';
import Moment from 'moment';

import './IssueList.scss';

const IssueList = ({ repos, issues, user, fetchIssues, fetchRepos, match, history }) => {
  useEffect(() => {
    if (!repos) {
      fetchRepos();
    }

    if (!issues) {
      fetchIssues(match.params.owner, match.params.name);
    }
  }, []);

  const avatarUrl = issue => issue.assignee ? issue.assignee.avatar_url : placeholderAvatar;

  const isActiveRepoClass = repo => repo.name === match.params.name ? 'IssueList__repos-row--active' : '';

  const changeRepo = repo => {
    history.push(`/repo/${repo.owner.login}/${repo.name}`);
    fetchIssues(repo.owner.login, repo.name);
  };

  const issueList = () => issues.map(issue =>(
    <div className="IssueList__issues-row" key={issue.id}>
      <span className="IssueList__issues-row--avatar">
        <img src={ avatarUrl(issue) } />
      </span>
      <span className="IssueList__issues-row--title">
        { issue.title }
      </span>
      <span className="IssueList__issues-row--created">
        { Moment(issue.created_at).format('MM/DD/YYYY') }
      </span>
      <span className="IssueList__issues-row--updated">
        { Moment(issue.updated_at).fromNow() }
      </span>
    </div>
  ));

  const repoList = () => repos.map(repo =>(
    <div className={`IssueList__repos-row ${isActiveRepoClass(repo)}`}
         onClick={() => changeRepo(repo)}
         key={repo.id}>
      <span className="IssueList__repos-row--name">
        { repo.name }
      </span>
    </div>
  ))

  return (
    <div className="IssueList">
      <div className="IssueList__repos">
      { repos && repoList() }
      </div>
      <div className="IssueList__issues">
        <div className="IssueList__issues-row IssueList__issues-row--header">
          <span className="IssueList__issues-row--avatar">
            Assignee
          </span>
          <span className="IssueList__issues-row--title">
            Issue Title
          </span>
          <span className="IssueList__issues-row--created">
            Created At
          </span>
          <span className="IssueList__issues-row--updated">
            Last Updated
          </span>
        </div>
        { issues && issueList() }
      </div>
    </div>
  );
}


export default IssueList;
