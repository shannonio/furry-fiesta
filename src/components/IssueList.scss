@import '../styles/core';

.IssueList {
  padding-top: 2rem;

  &__repos {
    display: grid;
    float: left;
    grid-template-rows: 1fr;
    padding: 1rem;
    align-items: start;
  }

  &__repos-row {
    display: grid;
    grid-template-columns: 1fr;
    justify-items: start;
    padding: 1rem;
    background-color: $light-grey;
    color: $white;

    &:hover, &__repos-row--active {
      font-weight: bold;
      cursor: pointer;
      background-color: $light-purple;
      transition: 0.218s all ease-in-out;
    }
  }

  &__issues {
    display: grid;
    grid-template-rows: 1fr;
  }

  &__issues-row {
    display: grid;
    grid-template-columns: 150px repeat(3, 1fr);
    justify-items: start;
    align-items: center;
    padding: .5rem 0;

    &:hover {
      cursor: pointer;
      background-color: $off-white;
    }
  }

  &__issues-row--header {
    font-weight: bold;
    background-color: $off-white;
    padding: 1rem 0;
  }

  &__issues-row--avatar {
    width: 100%;
  }

  &__issues-row--avatar img {
    height: 40px;
    border-radius: 50%;
  }
}
