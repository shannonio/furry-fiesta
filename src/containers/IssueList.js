import { connect } from 'react-redux';

import fetchIssues from '../actions/issues';
import fetchRepos from '../actions/repos';

import { getIssues } from '../reducers/issues';
import { getRepos } from '../reducers/repos';
import { getUser } from '../reducers/user';

import IssueList from '../components/IssueList'

const mapStateToProps = (state, ownProps) => {
  return {
    repos: getRepos(state),
    issues: getIssues(state),
    user: getUser(state)
  }
};

const mapDispatchToProps = (dispatch) => {
  return {
    fetchIssues: (username, repo) => fetchIssues(username, repo)(dispatch),
    fetchRepos: () => fetchRepos()(dispatch)
  }
};

const IssueListContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(IssueList);

export default IssueListContainer
