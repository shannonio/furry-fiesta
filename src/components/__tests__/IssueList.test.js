import React from 'react';
import ReactDOM from 'react-dom';
import { render, getByDisplayValue} from '@testing-library/react'

import IssueList from '../IssueList';

describe('IssueList', () => {
  const mockFetch = jest.fn();
  const repos = [];
  const issues = [];

  const createComponent = props => (
    render(
      <IssueList {...props} />
    )
  )

  it('renders without crashing', () => {
    const props = {
      repos,
      issues,
      fetchRepos: mockFetch,
      fetchIssues: mockFetch
    };

    const component = createComponent(props);

    expect(component).toMatchSnapshot();
  });

});
