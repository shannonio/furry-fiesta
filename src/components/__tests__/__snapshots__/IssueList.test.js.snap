// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssueList renders without crashing 1`] = `
<IssueList
  currentRepo={
    Object {
      "id": "1",
      "name": "Snarf",
      "owner": Object {
        "login": "Blaster",
      },
    }
  }
  fetchIssues={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "id": "Blaster",
          },
          "Snarf",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  fetchRepos={[MockFunction]}
  history={
    Object {
      "push": [MockFunction],
    }
  }
  issues={Array []}
  match={
    Object {
      "params": Object {
        "name": "Snarf",
        "owner": Object {
          "id": "Blaster",
        },
      },
    }
  }
  repos={
    Array [
      Object {
        "id": "1",
        "name": "Snarf",
        "owner": Object {
          "login": "Blaster",
        },
      },
      Object {
        "id": "2",
        "name": "Mermaid",
        "owner": Object {
          "login": "Little",
        },
      },
    ]
  }
  updatePrioritizedIssues={[MockFunction]}
>
  <div
    className="IssueList"
    onDrop={[Function]}
  >
    <div
      className="IssueList__row IssueList__row--header"
    >
      <span
        className="IssueList__row--avatar"
      >
        Assignee
      </span>
      <span
        className="IssueList__row--title"
      >
        Issue Title
      </span>
      <span
        className="IssueList__row--created"
      >
        Created At
      </span>
      <span
        className="IssueList__row--updated"
      >
        Last Updated
      </span>
    </div>
  </div>
</IssueList>
`;
